File: ios/SpentyAI/Core/Models/Transaction.swift
import Foundation

enum TransactionType: String, Codable, Hashable, CaseIterable {
    case income
    case expense
    case transfer
}

enum TransactionStatus: String, Codable, Hashable, CaseIterable {
    case approved
    case pendingReview = "pending_review"
    case rejected
}

struct Transaction: Codable, Identifiable, Hashable {
    var id: String { transactionId }

    let transactionId: String
    let userId: String?
    var transactionType: TransactionType
    var amount: Double
    var date: String
    var accountId: String?
    var toAccountId: String?
    var categoryId: String?
    var subcategoryId: String?
    var description: String?
    var paymentMethod: String?
    var isRecurring: Bool?
    var recurringFrequency: String?
    var source: String?
    var status: TransactionStatus?
    var receiptId: String?
    var createdAt: String?
    var accountName: String?
    var categoryName: String?

    enum CodingKeys: String, CodingKey {
        case transactionId = "transaction_id"
        case userId = "user_id"
        case transactionType = "transaction_type"
        case amount
        case date
        case accountId = "account_id"
        case toAccountId = "to_account_id"
        case categoryId = "category_id"
        case subcategoryId = "subcategory_id"
        case description
        case paymentMethod = "payment_method"
        case isRecurring = "is_recurring"
        case recurringFrequency = "recurring_frequency"
        case source
        case status
        case receiptId = "receipt_id"
        case createdAt = "created_at"
        case accountName = "account_name"
        case categoryName = "category_name"
    }
}

struct TransactionCreate: Codable {
    var transactionType: TransactionType
    var amount: Double
    var date: String
    var accountId: String?
    var toAccountId: String?
    var categoryId: String?
    var subcategoryId: String?
    var description: String?
    var paymentMethod: String?
    var isRecurring: Bool?
    var recurringFrequency: String?

    enum CodingKeys: String, CodingKey {
        case transactionType = "transaction_type"
        case amount
        case date
        case accountId = "account_id"
        case toAccountId = "to_account_id"
        case categoryId = "category_id"
        case subcategoryId = "subcategory_id"
        case description
        case paymentMethod = "payment_method"
        case isRecurring = "is_recurring"
        case recurringFrequency = "recurring_frequency"
    }
}

struct TransactionUpdate: Codable {
    var transactionType: TransactionType?
    var amount: Double?
    var date: String?
    var accountId: String?
    var toAccountId: String?
    var categoryId: String?
    var subcategoryId: String?
    var description: String?
    var paymentMethod: String?
    var isRecurring: Bool?
    var recurringFrequency: String?
    var status: TransactionStatus?

    enum CodingKeys: String, CodingKey {
        case transactionType = "transaction_type"
        case amount
        case date
        case accountId = "account_id"
        case toAccountId = "to_account_id"
        case categoryId = "category_id"
        case subcategoryId = "subcategory_id"
        case description
        case paymentMethod = "payment_method"
        case isRecurring = "is_recurring"
        case recurringFrequency = "recurring_frequency"
        case status
    }
}

File: android-native/app/src/main/java/com/spentyai/app/core/models/Transaction.kt
package com.spentyai.app.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Transaction(
    @SerialName("transactionId") val id: String = "",
    val transactionType: String? = null,
    val amount: Double? = null,
    val date: String? = null,
    val accountId: String? = null,
    val toAccountId: String? = null,
    val categoryId: String? = null,
    val subcategoryId: String? = null,
    val description: String? = null,
    val paymentMethod: String? = null,
    var status: String? = null,
    val isRecurring: Boolean? = null,
    val recurringFrequency: String? = null,
    val recurrenceDate: Int? = null,
    val source: String? = null,
    val receiptId: String? = null,
    val originalCurrency: String? = null,
    val originalAmount: Double? = null,
    val exchangeRate: Double? = null,
    val isEstimatedRate: Boolean? = null,
    @SerialName("source_email_id") val sourceEmailId: String? = null,
    @SerialName("source_sms_id") val sourceSmsId: String? = null,
    val categoryName: String? = null,
    var subcategoryName: String? = null,
    val accountName: String? = null,
    val toAccountName: String? = null
) {
    val sourceId: String? get() = sourceEmailId ?: sourceSmsId
}

@Serializable
data class TransactionListResponse(
    val transactions: List<Transaction> = emptyList(),
    val total: Int? = null
)

@Serializable
data class TransactionItemsResponse(
    val items: List<Transaction> = emptyList(),
    val total: Int? = null
) {
    fun asListResponse() = TransactionListResponse(transactions = items, total = total)
}

@Serializable
data class BulkIdsBody(
    @SerialName("transaction_ids") val transactionIds: List<String>
)

@Serializable
data class RecurringToggleBody(
    @SerialName("is_recurring") val isRecurring: Boolean,
    @SerialName("recurring_frequency") val recurringFrequency: String? = null
)

// Legacy enum kept for backward compatibility but the API uses string types
@Serializable
enum class TransactionType {
    @SerialName("income") INCOME,
    @SerialName("expense") EXPENSE,
    @SerialName("transfer") TRANSFER,
    @SerialName("refund") REFUND,
    @SerialName("adjustment") ADJUSTMENT
}
